src/controller/smartapi.py

```diff
@@ -0,0 +1,182 @@
+"""
+    SmartAPIEntity CRUD Validation and Refresh Operations
+
+    Validation only:
+
+        smartapi = SmartAPIEntity(SmartAPIEntity.VALIDATION_ONLY)
+        smartapi.raw = rawbytes
+        smartapi.validate()
+
+    Add a document:
+
+        smartapi = SmartAPIEntity(url)
+        smartapi.raw = rawbytes
+
+        smartapi.username = username
+        smartapi.slug = slug # optional
+
+        smartapi.validate() # should be called before saving
+        smartapi.save()
+
+        smartapi.check() # populate uptime status
+        smartapi.refresh() # refresh and populate refresh status, auto validate
+        smartapi.save()
+
+    Modify a document metadata:
+
+        smartapi = SmartAPIEntity.get(_id)
+        smartapi.slug = newslug
+        smartapi.save()
+
+    Delete a document:
+
+        smartapi = SmartAPIEntity.get(_id)
+        smartapi.delete()
+
+"""
+import logging
+import string
+from collections.abc import Mapping
+from datetime import datetime, timezone
+from warnings import warn
+
+from model import SmartAPIDoc
+from .base import AbstractWebEntity, OpenAPI, Swagger
+from .exceptions import ConflictError, ControllerError
+
+logger = logging.getLogger(__name__)
+
+# TODO etag support
+
+
+class Slug:
+    """
+    Optional secondary key for a DB entry.
+    It must be a unique string or None.
+    """
+
+    def __get__(self, obj, objtype=None):
+        if obj is None:  # class atrribute
+            return self
+        return getattr(obj, "_slug", None)
+
+    def __set__(self, obj, value):
+        self.validate(value)
+        setattr(obj, "_slug", value)
+
+    def validate(self, value):
+
+        if value is None:
+            return  # this field is optional, okay to be None
+
+        if not isinstance(value, str) or len(value) < 3:
+            raise ValueError("Slug must be at least 3 characters.")
+
+        if value != value.lower():
+            raise ValueError("Slug must be in lowercase.")
+
+        if value in ("www", "dev", "smart-api"):
+            raise ValueError(f"Slug '{value}' not allowed.")
+
+        _valid_chars = string.ascii_letters + string.digits + "-_"
+        if not all(c in _valid_chars for c in value):
+            raise ValueError("Slug contains invalid characters.")
+
+
+class SmartAPIEntity(AbstractWebEntity, Mapping):
```

Comment:
Why is it renamed from `SmartAPI` to `SmartAPIEntity`? Seems not necessary to me @sengineer0

Reply:
After I added a new model to store MetaKG, we have controllers, models for MetaKG, SmartAPI. So I think it will make the code more readable and we can easily determine what obj type is.
@newgene